websocketOutportDriver/websocketOutportDriverNodePart.go

```diff
@@ -0,0 +1,202 @@
+		RoundsInfos: roundsInfos,
+	}
+	err := o.handleAction(args, outportSenderData.OperationSaveRoundsInfo)
+	if err != nil {
+		return fmt.Errorf("%w on SaveRoundsInfo", err)
+	}
+
+	return nil
+}
+
+// SaveValidatorsPubKeys will handle the saving of the validators' public keys
+func (o *websocketOutportDriverNodePart) SaveValidatorsPubKeys(validatorsPubKeys map[uint32][][]byte, epoch uint32) error {
+	args := outportSenderData.ArgsSaveValidatorsPubKeys{
+		ValidatorsPubKeys: validatorsPubKeys,
+		Epoch:             epoch,
+	}
+	err := o.handleAction(args, outportSenderData.OperationSaveValidatorsPubKeys)
+	if err != nil {
+		return fmt.Errorf("%w on SaveValidatorsPubKeys", err)
+	}
+
+	return nil
+}
+
+// SaveValidatorRating will handle the saving of the validators' rating
+func (o *websocketOutportDriverNodePart) SaveValidatorsRating(indexID string, infoRating []*indexer.ValidatorRatingInfo) error {
+	args := outportSenderData.ArgsSaveValidatorsRating{
+		IndexID:    indexID,
+		InfoRating: infoRating,
+	}
+	err := o.handleAction(args, outportSenderData.OperationSaveValidatorsRating)
+	if err != nil {
+		return fmt.Errorf("%w on SaveValidatorsRating", err)
+	}
+
+	return nil
+}
+
+// SaveAccounts will handle the accounts' saving
+func (o *websocketOutportDriverNodePart) SaveAccounts(blockTimestamp uint64, acc []data.UserAccountHandler) error {
+	args := outportSenderData.ArgsSaveAccounts{
+		BlockTimestamp: blockTimestamp,
+		Acc:            acc,
+	}
+	err := o.handleAction(args, outportSenderData.OperationSaveAccounts)
+	if err != nil {
+		return fmt.Errorf("%w on SaveAccounts", err)
+	}
+
+	return nil
+}
+
+// FinalizedBlock will handle the finalized block
+func (o *websocketOutportDriverNodePart) FinalizedBlock(headerHash []byte) error {
+	args := outportSenderData.ArgsFinalizedBlock{
+		HeaderHash: headerHash,
+	}
+	err := o.handleAction(args, outportSenderData.OperationFinalizedBlock)
+	if err != nil {
+		return fmt.Errorf("%w on FinalizedBlock", err)
+	}
+
+	return nil
+}
+
+// Close will handle the closing of the outport driver web socket sender
+func (o *websocketOutportDriverNodePart) Close() error {
+	// TODO: close the web socket here
+	return nil
+}
+
+// IsInterfaceNil returns true if there is no value under the interface
+func (o *websocketOutportDriverNodePart) IsInterfaceNil() bool {
+	return o == nil
+}
+
+func (o *websocketOutportDriverNodePart) handleAction(args interface{}, operation outportSenderData.OperationType) error {
+	marshaledBlock, err := o.marshalizer.Marshal(args)
+	if err != nil {
+		o.log.Error("cannot marshal block", "operation", operation.String(), "error", err)
+		return err
+	}
+
+	payload := o.preparePayload(operation, marshaledBlock)
+
+	for {
+		err = o.webSocketSender.SendOnRoute(outportSenderData.WsSendArgs{
+			Payload:   payload,
+			Operation: operation,
+		})
+		if err != nil {
+			o.log.Error("cannot send on route. Retrying", "operation", operation.String(), "error", err)
```

Comment:
didn't we have a retrial mechanism already in the webSocketSender implementation?

Reply:
fixed